### ISSUE TITLE
jsdom loading issue if threads false

### ISSUE THREAD

**Issue by bkuster** [vitest-dev/vitest#835]
### Describe the bug

# Issue
When starting vitest with `threads: false` tests fail with: `Cannot read properties of undefined (reading 'child')` The same test run while using `threads: true`.

# Possible Causes
The test uses the @vue/test-utils package for component testing. These _must_ be loaded after jsdom is loaded and the DOM is provided. (see [this issue](https://github.com/vuejs/vue-test-utils/issues/1192)).

# Failed workarounds
I tried loading jsdom via jsdom-global myself in a global setup hook, so far to no avail.

### Reproduction

# vite.config.js
```js
import { defineConfig } from 'vite';
import { createVuePlugin } from 'vite-plugin-vue2';

const configMain = defineConfig({
  plugins: [
    createVuePlugin(),
  ],
  test: {
    threads: false,
    environment: 'jsdom',
  },
});

export default configMain;
```

# Component.spec.js
```js
import { describe, expect, it } from 'vitest';
import { mount, createLocalVue } from '@vue/test-utils';
import Component from './Component.vue';

describe('Compontent', () => {
  const localVue = createLocalVue();

  it('mounts', () => {
    const wrapper = mount(Component, { localVue });
    expect(wrapper).toMatchSnapshot();
  });
});
```
# Component.vue
```vue
<template>
  <div>
    Foo Bar
  </div>
</template>

<script>
  export default {
    name: 'FooBar',
  };
</script>
```
### setup
```bash
npm i vitest vite vue@^2.6.0 @vue/test-utils@1 jsdom
```

### System Info

```shell
System:
    OS: Linux 5.4 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (6) x64 Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
    Memory: 643.27 MB / 31.36 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
  Browsers:
    Chromium: 98.0.4758.102
    Firefox: 97.0
  npmPackages:
    vite: ^2.8.4 => 2.8.4 
    vitest: ^0.5.3 => 0.5.3
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by cexbrayat** [comment]
I tried to dig a bit more into this issue, and it turns out that the teardown function of the jsdom environment is never called when running with threads https://github.com/vitest-dev/vitest/blob/cf226c5a0eb98bf3c9fd5f97d8087a30e0daa7a0/packages/vitest/src/integrations/env/jsdom.ts#L61-L63
This is because the worker pool never finishes and the `finally` here is never called
https://github.com/vitest-dev/vitest/blob/cf226c5a0eb98bf3c9fd5f97d8087a30e0daa7a0/packages/vitest/src/runtime/setup.ts#L148-L154

It is only called with threads disabled (when there is no pool), and that leads to inconsistencies between the two modes.
It is highly possible that the teardown function is not correct, but that the issue was never spotted because the CI tests of this repo run with threaded. I added a simple repro in #1060 

The same probably happens with the other environments like happy-dom as well.
The repro I added also surfaces some mock tests failing, but I haven't found an obvious reason.

TL;DR:
- the jsdom environment teardown function is only called when no threads are used, whereas it should be run even with threads
- the jsdom environment teardown function probably needs to be fixed

**Comment by sheremet-va** [comment]
I think one of the problems is tests using the same `@vue/test-utils` module, which tries to read inaccessible jsdom instances. I tried fixing it in #1102 - let me know if it at least partially fixes the issue?

Also, I agree, teardown function may need improvement.

**Comment by sheremet-va** [comment]
Also, I want to point out that Vitest treats workers differently than Jest. Jest provides isolation with `vm` module, but we cannot do that due to poor ESM support, so for isolation we are heavily relying on isolated workers. Disabled workers you are disabling Vitest's isolation and have to do it yourself.

Please, don't expect `--no-threads` to act the same as `--runInBand`.

**Comment by cexbrayat** [comment]
@sheremet-va Thanks for your feedback.

Yeah, I was trying to fix CI issues that occur since vitest 0.7+, and thought that `--no-threads` could save the day (as `--runInBand` was sometimes helpful on CI machines for Jest), but I now understand the concepts are not exactly the same. That could be worth documenting, as I suppose I'm not the only one who we'll make a similar assumption. 

If I understand correctly #1102, we would need to call `vi.resetModules()` before each test? If so, would it be possible to have a config option as we do for `clearMocks` for example?

**Comment by sheremet-va** [comment]
> If I understand correctly https://github.com/vitest-dev/vitest/pull/1102, we would need to call vi.resetModules() before each test? If so, would it be possible to have a config option as we do for clearMocks for example?

It is called automatically. I've tried calling it in your repro, and came to this conclusion:

I found that vue uses cached `runtime-dom.cjs.js`, removing it from cache helped: 

```
Object.keys(require.cache)
  .filter(k => k.includes('vue'))
  .forEach(k => delete require.cache[k])
  
import { mount } from '@vue/test-utils'
```

I guess what happens is vue plugin imports some instance of vue, which gets cached and is returned every time. It's fine if it was `required`, but we can't clear cache if it was imported (since Node doesn't expose `import` cache), but we haven't encountered problems with it yet.

**Comment by sheremet-va** [comment]
So, to paint a full picture:

1. Running `vitest --no-threads`
2. Vitest starts a Vite server
3. Vite plugins are initialized
3.5. Setup jsdom/happy-dom
4. Vitest starts tests in the SAME thread (which means it shares a state with Vite server and all it's caches)
5. Test imports `@vue/test-utils`
6. `@vue/test-utils` imports cached on `require.cache` instance of `vue`, which was initialized BEFORE jsdom is setup
7. Fail

**Comment by sheremet-va** [comment]
I have tried clearing the cache entirely and/or depending on what was required in `plugins`, but this cases other problems.

So, as a patch solution, I propose `deps.clear` config options: an array of regexps to clear from cache before each test. This should help, if a package depends on a global variable, defined when requiring (like Vue does).

**Comment by aml3ed** [comment]
Any news on this @antfu @patak-dev @sheremet-va? I'm experiencing a similar problem using `--no-threads` and `@testing-library/react`. First run passes, but watched second run fails.

Let me know if I can provide any additional information 👍

**Comment by sheremet-va** [comment]
So, with `--no-threads` we should define environment only one time. If we remove `teardown` and don't run `new JSDOM`, when window is already defined, all tests are passing.

This is usually ok, if you don't use `@vitest-environment`, because it shutters this logic a bit. So, I propose we define window only once and throw an error, if someone uses different `@vitest-environment` for now. To give people time to test this approach. After that we can try to figure out other usecase.

For example, we can group files by different environments.

**Comment by sheremet-va** [comment]
I also want to point out that I don't recommend using `--no-threads` with Vue and React projects, because you loos isolation, and have to provided it yourself. 

I guess usually, if you don't rely on some global variable, and cleanup/unmount after each test, it should be ok. But disable `--threads` on your own risk!
